File: haskell-server/test/BoardGame/Server/Domain/WordDictionarySpec.hs
--
-- Copyright 2017 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}

module BoardGame.Server.Domain.WordDictionarySpec where

import Test.Hspec

import qualified Data.List as List
import qualified Data.Set as Set

import Bolour.Language.Domain.WordDictionary (WordDictionary, WordDictionary(WordDictionary))
import qualified Bolour.Language.Domain.WordDictionary as Dict

maxMaskedWords :: Int
maxMaskedWords = 2
myWords :: [String]
myWords = ["GLASS", "TABLE", "SCREEN", "NO", "ON"]
maskedWords = Set.toList $ Dict.mkMaskedWords myWords maxMaskedWords
dictionary = Dict.mkDictionary "en" myWords maskedWords maxMaskedWords

spec :: Spec
spec = do
  describe "test finding word in dictionary" $ do
    it "check existing word" $ do
      Dict.isWord dictionary "GLASS" `shouldBe` True
    it "check non-existent word" $ do
      Dict.isWord dictionary "GLAS" `shouldBe` False
  describe "test finding word permutations" $ do
    it "find existing word permutation" $ do
      Dict.getWordPermutations dictionary "ABELT" `shouldBe` ["TABLE"]
    it "no word permutations" $ do
      Dict.getWordPermutations dictionary "ABEL" `shouldBe` []
    it "2 word permutations" $ do
      (List.sort $ Dict.getWordPermutations dictionary "NO") `shouldBe` ["NO", "ON"]

File: scala-server/test/com/bolour/boardgame/scala/server/domain/WordDictionarySpec.scala
/*
 * Copyright 2017-2018 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
 */

package com.bolour.boardgame.scala.server.domain

import com.bolour.boardgame.scala.server.util.WordUtil
import com.bolour.language.scala.domain.WordDictionary
import org.scalatest.{FlatSpec, Matchers}
import com.bolour.language.scala.domain.WordDictionary._
import org.slf4j.LoggerFactory

class WordDictionarySpec extends FlatSpec with Matchers {
  val logger = LoggerFactory.getLogger(this.getClass)

  "mask words" should "be computed" in {
    val words = List("FOX", "RAT", "BIRD")
    val maskedWords1 = mkMaskedWords(words, 1)
    maskedWords1 should contain ("F X")
    maskedWords1 should not contain ("R  ")

    val maskedWords2 = mkMaskedWords(words, 2)
    maskedWords2 should contain ("F X")
    maskedWords2 should contain ("R  ")

    println(maskedWords2)
  }

  "masked words" should "be read" in {
    val dictionary = WordDictionary.mkWordDictionary("tiny", "dict", 3).get
    println(dictionary.maskedWords)
    dictionary.hasMaskedWord("N T") shouldBe true
  }
}
